File: src/main/java/de/hsa/bpmnandai/project/domain/facade/SimpleDmnFacade.java
package de.hsa.bpmnandai.project.domain.facade;

import de.hsa.bpmnandai.project.domain.service.CreditService;
import de.hsa.bpmnandai.project.domain.service.ProcessService;
import de.hsa.bpmnandai.project.infrastructure.entity.BankLoanRequestEntity;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.camunda.bpm.engine.history.HistoricProcessInstance;
import org.springframework.stereotype.Component;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.concurrent.atomic.AtomicInteger;

@Slf4j
@Component
@RequiredArgsConstructor
public class SimpleDmnFacade {

    private final ProcessService processService;
    private final CreditService creditService;

    private final String PROCESS_KEY = "Process_KreditAnfrage";

    public void processLoanRequests() {
        final List<BankLoanRequestEntity> entities = this.creditService.getAllRequests();
        final int totalEntries = entities.size();
        int currentIteration = 0;
        log.info("Number of loan-requests: {}", entities.size());
        for (final BankLoanRequestEntity request : entities) {
            final Map<String, Object> processVariables = getProcessVariables(request);
            final String businessKey = request.getId();
            processService.startInstanceOfProcess(PROCESS_KEY, businessKey, processVariables);
            currentIteration++;
            log.info("Provided loan-request '{}' of '{}' to engine", currentIteration, totalEntries);
        }
    }

    public void addHistory() {
        final List<BankLoanRequestEntity> entities = this.creditService.getAllRequests();
        final AtomicInteger iteration = new AtomicInteger();
        final List<HistoricProcessInstance> historic = processService.getAllHistoricProcessInstances();
        historic.forEach(historicInstance -> {
            final String businessKey = historicInstance.getBusinessKey();
            final Optional<BankLoanRequestEntity> entity = entities.stream().filter(e -> e.getId().equals(businessKey)).findFirst();
            entity.ifPresent(e -> e.setPredictionIsApproved(historicInstance.getEndActivityId()));
            log.info("Requested result nr. '{}'", iteration.incrementAndGet());
        });
        this.creditService.saveCreditRequests(entities);
    }

    private Map<String, Object> getProcessVariables(final BankLoanRequestEntity entity) {
        final Map<String, Object> varMap = new HashMap<>();
        varMap.put("emi", entity.getEmi());
        varMap.put("city_category", entity.getCityCategory());
        varMap.put("loan_period", entity.getLoanPeriod());
        varMap.put("monthly_income", entity.getMonthlyIncome());
        varMap.put("interest_rate", entity.getInterestRate());
        varMap.put("employer_category1", entity.getEmployerCategory());
        varMap.put("source_category", entity.getSourceCategory());
        varMap.put("gender", entity.getGender());
        varMap.put("existing_emi", entity.getExistingEmi());
        varMap.put("loan_amount", entity.getLoanAmount());
        varMap.put("contacted", entity.getContacted());
        return varMap;
    }

}

File: src/main/kotlin/de/emaarco/bpmnandai/project/domain/facade/SimpleDmnFacade.kt
package de.emaarco.bpmnandai.project.domain.facade

import de.emaarco.bpmnandai.project.domain.service.CreditService
import de.emaarco.bpmnandai.project.domain.service.ProcessService
import de.emaarco.bpmnandai.project.infrastructure.entity.BankLoanRequestEntity
import org.camunda.bpm.engine.history.HistoricProcessInstance
import org.springframework.stereotype.Component

@Component
class SimpleDmnFacade(
    private val processService: ProcessService,
    private var creditService: CreditService,
) {

    private val PROCESS_KEY = "Process_KreditAnfrage"

    fun processLoanRequests() {
        val entities: List<BankLoanRequestEntity> = creditService.getAllRequests()
        val totalEntries = entities.size
        var currentIteration = 0
        for (request in entities) {
            val processVariables = getProcessVariables(request)
            val businessKey: String = request.requestId
            processService.startInstanceOfProcess(PROCESS_KEY, businessKey, processVariables)
            currentIteration++
            println("Provided loan-request '${currentIteration}' of '${totalEntries}' to engine")
        }
    }

    fun addHistory() {
        val entities: List<BankLoanRequestEntity> = creditService.getAllRequests()
        var iteration = 0;
        val historic: List<HistoricProcessInstance> = processService.getAllHistoricProcessInstances()
        historic.forEach { instance ->
            val entity: BankLoanRequestEntity? = entities.find { e -> e.requestId == instance.businessKey };
            entity?.predictionIsApproved = instance.endActivityId
            println("Requested result nr. '${++iteration}'")
        };
        creditService.saveCreditRequests(entities)
    }

    private fun getProcessVariables(entity: BankLoanRequestEntity): Map<String, Any?> {
        val varMap: MutableMap<String, Any?> = HashMap()
        varMap["emi"] = entity.emi
        varMap["city_category"] = entity.cityCategory
        varMap["loan_period"] = entity.loanPeriod
        varMap["monthly_income"] = entity.monthlyIncome
        varMap["interest_rate"] = entity.interestRate
        varMap["employer_category1"] = entity.employerCategory
        varMap["source_category"] = entity.sourceCategory
        varMap["gender"] = entity.gender
        varMap["existing_emi"] = entity.existingEmi
        varMap["loan_amount"] = entity.loanAmount
        varMap["contacted"] = entity.contacted
        return varMap
    }

}